### PR TITLE
tracing: store list of contacted replicas in otel tracing

### PR DESCRIPTION
**Pull request by margdoc**
Add information about contacted replicas to opentelemetry_state.

### CODE REVIEW

File: service/storage_proxy.cc [hunk 1]
```diff
@@ -1456,6 +1456,10 @@ storage_proxy::response_id_type storage_proxy::create_write_response_handler(key
     shared_ptr<abstract_write_response_handler> h;
     auto& rs = ks.get_replication_strategy();
 
+    if (tr_state.has_opentelemetry()) {
+        tr_state.get_opentelemetry_ptr()->set_replicas(targets);
+    }
```
Comment (constantfold) [inline service/storage_proxy.cc]: Maybe similarly to other methods in `trace_state.hh` make a helper method for `set_replicas`?

(see `set_username` in `trace_state.hh`)
```cpp
inline void set_username(const trace_state_ptr& p, const std::optional<auth::authenticated_user>& user) {
    if (p.has_tracing()) {
        p.get_tracing_ptr()->set_username(user);
    }
}
```

File: tracing/trace_state.hh [hunk 2]
```diff
@@ -510,6 +511,24 @@ class opentelemetry_state final {
 private:
     lw_shared_ptr<trace_state> _state_ptr;
     bool const _opentelemetry_tracing{false};
+    inet_address_vector_replica_set _replicas;
+
+    void serialize_replicas(bytes& serialized) const noexcept {
```
Comment (constantfold) [inline tracing/trace_state.hh]: Please move the implementation to `trace_state.cc`

Comment (constantfold) [reply tracing/trace_state.hh]: This function should not be marked `noexcept` as appending to `bytes` allocates memory which can fail and throw exception.

Comment (jul-stas) [reply tracing/trace_state.hh]: Similar thing with `noexcept` in `set_replicas()`, although that one can be mitigated by accepting a copy and doing a move-assignment (which is `noexcept` for `small_vector`)

**Comment by jul-stas** [comment]
LGTM (modulo the remarks above)

**Pull request merged**